librz/util/thread.c

```diff
@@ -24,45 +26,43 @@
 #include <OS.h>
 #endif
 
-#if __WINDOWS__
-static DWORD WINAPI _rz_th_launcher(void *_th) {
-#else
-static void *_rz_th_launcher(void *_th) {
+struct rz_th_pool_t {
+	size_t size;
+	RzThread **threads;
+};
+
+struct rz_th_queue_t {
+	RzThreadLock *lock;
+	RzThreadCond *cond;
+	size_t max_size;
+	RzList *list;
+};
+
+/*
+ * Main thread function, this function is meant to be
+ * hidden from the user which is using the C APIs.
+ */
+static RZ_TH_RET_T thread_main_function(void *_th) {
+#if HAVE_PTHREAD
+#ifndef __ANDROID__
+	pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, NULL);
+	pthread_setcanceltype(PTHREAD_CANCEL_ASYNCHRONOUS, NULL);
 #endif
-	int ret;
-	RzThread *th = _th;
-	th->ready = true;
-	if (th->delay > 0) {
-		rz_sys_sleep(th->delay);
-	} else if (th->delay < 0) {
-		rz_th_lock_wait(th->lock);
-	}
-	rz_th_lock_enter(th->lock);
-	do {
-		rz_th_lock_leave(th->lock);
-		th->running = true;
-		ret = th->fun(th);
-		if (ret < 0) {
-			// th has been freed
-			return 0;
-		}
-		th->running = false;
-		rz_th_lock_enter(th->lock);
-	} while (ret);
-	rz_th_lock_leave(th->lock);
+#endif
+
+	RzThread *th = (RzThread *)_th;
+	RzThreadStatus status = RZ_TH_STATUS_LOOP;
+
+	while (status == RZ_TH_STATUS_LOOP) {
```

Comment:
The looping-"feature" should be removed from `RzThread` too since it has absolutely no benefits over just putting a while loop in `th->function`. On the contrary actually: by removing it, we will de-occupy the return value of `th->function` to be used by the user of `RzThread` for any purpose.

Reply:
i agree/disagree. thread functions for me should be returning `void`.
i did not remove it because it required to rewrite the code related to the rtr server which i do not want to handle now.

Reply:
Ok, if it's not an easy change then it's ok for me to leave in for now. But please maybe mark with RZ_DEPRECATED where appropriate.

Reply:
i think i will probably just rewrite the code, to make everybody happy.
but just to be sure, do we agree to return `void` and make the API consistent? or i need to return a pointer and add the return value to the structure?

Reply:
Imo `void *`. But technically it does not have to be added to the structure but can instead be returned directly when doing the join, see for example https://git.sr.ht/~thestr4ng3r/chiaki/tree/420809b24ef7bb8510a13d9a4c02ac260a9841a7/item/lib/src/thread.c#L59-73

Reply:
Interestingly, libuv, which provides similar platform-agnostic wrappers, does not use the native return values in `uv_thread_join()`: http://docs.libuv.org/en/v1.x/guide/threads.html#core-thread-operations

Reply:
i have added `rz_th_retv` because the join one is not compatible on how the pool code works (i.e. a user will never be able to retrieve the return value).
i also changed it to `void *` as return type.

Reply:
I see what you mean. However I think you should not store the return value **inside** the thread, but rather return it from the outer thread function and retrieve it inside the join and then store away in the struct there, to avoid any potential synchronization issues and enforce that the thread has been joined when the return value is accessed from the outside.
From what I understand, the `pthread_exit()` can also be removed completely since it is useless when there is no code after it (like calling `exit(0)` at the end of a main function).

Reply:
- regarding pthread_exit, you are right, we can avoid calling it.
- regarding where to store the return value, i disagree. that value is meant to not be accessed in a thread-safe way. even in your code the issue is the same on windows side. i think we can keep it in this way, otherwise i need to add something like a RzList just to get all return values..

Reply:
Oh you are perfectly right, I read my own code wrong there. Then it's fine and the caller has to make sure to join before accessing it.